currencies/rate_converter.go

```diff
@@ -0,0 +1,165 @@
+
+	updatedRates := &Rates{}
+	err = json.NewDecoder(response.Body).Decode(&updatedRates)
+	if err != nil {
+		return nil, err
+	}
+	updatedRates.ComputeDataAsOf()
+
+	return updatedRates, err
+}
+
+// updateRates allows to update the internal currencies rates map with the one provided
+func (rc *RateConverter) updateRates(updatedRates *Rates) {
+	rc.Lock()
+	defer rc.Unlock()
+	rc.rates = updatedRates
+	now := time.Now()
+	rc.lastUpdated = &now
+}
+
+// Update updates the internal currencies rates from remote sources
+func (rc *RateConverter) Update() (*Rates, error) {
+	rates, err := rc.fetch()
+	if err == nil {
+		rc.Lock()
+		defer rc.Unlock()
+		rc.rates = rates
+		now := time.Now()
+		rc.lastUpdated = &now
+	}
+
+	return rates, err
+}
+
+// StartPeriodicFetching starts the periodic fetching at the given interval
+// triggers a first fetch when called before the first tick happen in order to initialize currencies rates map
+// returns a chan in which the number of data updates everytime a new update was done
+func (rc *RateConverter) StartPeriodicFetching() (chan int, error) {
+	rc.ticker = time.NewTicker(rc.fetchingInterval)
+	updatesTicksCount := 0
+	updatesNotifier := make(chan int)
+
+	go func() {
+		for {
+			select {
+			case <-rc.ticker.C:
+				go func() {
+					rates, err := rc.fetch()
+					if err == nil {
+						rc.updates <- rates
+					}
+					// TODO: Handle retry strategy
+				}()
+			case rates := <-rc.updates:
+				go func() {
+					rc.updateRates(rates)
+					updatesTicksCount++
+					updatesNotifier <- updatesTicksCount
+				}()
+			case <-rc.done:
+				if rc.ticker != nil {
+					rc.ticker.Stop()
+					rc.ticker = nil
+				}
+				return
+			default:
+			}
+		}
+	}()
+
+	// Trigger a first tick at the very begining
+	// the ticker will tick later
+	rates, err := rc.fetch()
+	if err != nil {
+		return nil, err
+	}
+	rc.updateRates(rates)
+
+	return updatesNotifier, nil
+}
+
+// StopPeriodicFetching stops the periodic fetching while keeping the latest currencies rates map
+func (rc *RateConverter) StopPeriodicFetching() {
+	rc.done <- true
+}
+
+// LastUpdated returns time when currencies rates were updated
+func (rc *RateConverter) LastUpdated() *time.Time {
+	return rc.lastUpdated
+}
+
+// Rates returns current rates
+func (rc *RateConverter) Rates() *Rates {
+	return rc.rates
+}
+
+// GetRate returns the conversion rate between two currencies
+// returns an error in case the conversion rate between the two given currencies is not in the currencies rates map
+func (rc *RateConverter) GetRate(from string, to string) (*float32, error) {
+	if rc.rates != nil {
+		rc.RLock()
+		defer rc.RUnlock()
+		if conversion, present := rc.rates.Conversions[from][to]; present == true {
+			return &conversion, nil
+		}
+		return nil, fmt.Errorf("conversion %s->%s not present in rates dictionnary", from, to)
+	}
+	return nil, errors.New("rates are nil")
+}
+
+type httpClient interface {
+	Get(url string) (*http.Response, error)
```

Comment:
Nice use of interface! I only recently learned how awesome it is that you can define ad-hoc interfaces like this wherever you need them, and I feel like it was a major step forward in my Go development skills :).

One issue I see in this PR is that there's nothing cancelling the request if `currency.prebid.org` hangs for some reason. This also affects  your `TODO: Handle retry strategy`, since you'll want to cancel the first request when you do a retry.

All that said: I think `Do(req *Request) (*Response, error)` might make a better interface function here. You need an `http.Request` if you want to [attach Contexts](https://golang.org/pkg/net/http/#Request.WithContext), which you'll need to cancel requests before you retry them.

Reply:
I'll have a look yes, good point.

Reply:
Done